File: day010_string_parsing/related_valid_parentheses/solution.cpp
#include <iostream>
#include <string>
#include <stack>
#include <unordered_map>

using namespace std;

class Solution {
public:
    // Optimized: O(N) Time, O(N) Space - Stack and Hash Map
    bool isValid_optimal(string s) {
        // Hash Map for O(1) lookup: closer -> opener
        unordered_map<char, char> mappings = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        stack<char> bracket_stack;

        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                // Case 1: Opener -> Push to stack
                bracket_stack.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                // Case 2: Closer
                
                // If stack is empty (unmatched closer) or mismatch
                if (bracket_stack.empty() || bracket_stack.top() != mappings[c]) {
                    return false;
                }
                // Match found -> Pop the opener
                bracket_stack.pop();
            }
        }
        
        // Final check: The string is valid only if the stack is completely empty
        return bracket_stack.empty();
    }

    // Brute Force: O(N) Time - The stack approach is the standard linear method.
    bool isValid_brute(string s) {
        return isValid_optimal(s);
    }

    void solve() {
        string s;
        if (!getline(cin, s)) return;
        
        if (isValid_optimal(s)) {
            cout << "true" << endl;
        } else {
            cout << "false" << endl;
        }
    }
};

int main() {
    Solution().solve();
    return 0;
}

File: day010_string_parsing/related_valid_parentheses/solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.Stack;

class Solution {

    // Optimized: O(N) Time, O(N) Space - Stack and Hash Map
    public boolean isValid_optimal(String s) {
        // Hash Map for O(1) lookup: closer -> opener
        Map<Character, Character> mappings = new HashMap<>();
        mappings.put(')', '(');
        mappings.put('}', '{');
        mappings.put(']', '[');
        
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (mappings.containsValue(c)) {
                // Case 1: Opener -> Push to stack
                stack.push(c);
            } else if (mappings.containsKey(c)) {
                // Case 2: Closer
                
                // If stack is empty (unmatched closer) or mismatch
                if (stack.isEmpty() || stack.pop() != mappings.get(c)) {
                    return false;
                }
            }
            // Ignore other characters
        }
        
        // Final check: Valid only if the stack is empty
        return stack.isEmpty();
    }

    // Brute Force: O(N) Time - The stack approach is the standard linear method.
    public boolean isValid_brute(String s) {
        return isValid_optimal(s);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String s = scanner.nextLine();
        
        Solution solution = new Solution();
        if (solution.isValid_optimal(s)) {
            System.out.println("true");
        } else {
            System.out.println("false");
        }
        scanner.close();
    }
}